### PR TITLE
API for running script on an image

### PR DESCRIPTION
**Pull request by djfinn14**
This API allows SecCloud to run a custom script on an image. BMI is responsible for mapping and unmapping the image.

### CODE REVIEW

File: ims/common/constants.py [hunk 1]
```diff
@@ -124,4 +126,4 @@
 
 BMI_ADMIN_PROJECT = "bmi_infra"
 
-HAAS_BMI_CHANNEL = "vlan/native"
+HAAS_BMI_CHANNEL = "null"
```
Comment (naved001) [inline ims/common/constants.py]: why did you change this to null? that wouldn't work in real life.

File: ims/cli/cli.py [hunk 2]
```diff
@@ -159,7 +161,7 @@ def remove_image(project, img):
 
 @cli.command(name='ls', short_help='List Images Stored')
 @click.argument(constants.PROJECT_PARAMETER)
-def list_images(project):
+def list_image(project):
```
Comment (naved001) [inline ims/cli/cli.py]: why change of name here?

File: ims/einstein/operations.py [hunk 3]
```diff
@@ -635,3 +639,31 @@ def remake_mappings(self):
...
+    def run_script(self, img, script):
+        try:
+            if not self.is_admin:
+                raise AuthorizationFailedException()
+            directory = "/tmp/script/"
+            os.makedirs(directory)
+            script_path = directory + "/test_script.sh"
+            ret = self.fs.map(img)
+            decode_script = base64.b64decode(script)
+            with open(script_path, "w") as write_script:
```
Comment (chemistry-sourabh) [inline ims/einstein/operations.py]: Shouldn't this be wb ? as you did a rb when reading

File: ims/einstein/operations.py [hunk 4]
```diff
@@ -635,3 +639,31 @@ def remake_mappings(self):
...
+                write_script.write(decode_script)
+            os.chmod(script_path, S_IRWXU)
+            p = subprocess.Popen(script_path + " " + ret, shell=True,
```
Comment (chemistry-sourabh) [inline ims/einstein/operations.py]: This shell=True is dangerous. Also we are just mapping the image, but not mounting it on the VM due to which the script will not be able to access the contents of the image.

The right way to do it would be to mount the image using the "mount" command in some location in /tmp, chroot to that location and then run the script.

Right now if the user is admin he can technically run any script which can be "rm -rf /" which is dangerous. The chrooting will help with this, but might delete the entire image. So it would be better to clone the image before mapping and deleting it at the end.

File: ims/einstein/operations.py [hunk 5]
```diff
@@ -635,3 +639,31 @@ def remake_mappings(self):
...
+                                 stderr=subprocess.STDOUT,
+                                 stdout=subprocess.PIPE)
+            output, err = p.communicate()
+            code = p.returncode
+            encode_output = base64.b64encode(output)
+            return self.__return_success({'stdout': encode_output,
```
Comment (chemistry-sourabh) [inline ims/einstein/operations.py]: There is no need to send err as it will be included in stdout.

File: ims/einstein/operations.py (hunk 4 above)
Comment (chemistry-sourabh) [inline ims/einstein/operations.py]: Also PS mount will not work for all filesystems. You will probably need to do more stuff to support those filesystems.

Comment (djfinn14) [inline ims/einstein/operations.py]: The design we decided on with the Secure Cloud team has their script performing the actual mounting of the image. I will change it to clone the image before mapping and delete it after.

File: ims/cli/cli.py [hunk 6]
```diff
@@ -671,5 +673,47 @@ def download():
     click.echo('Not Yet Implemented')
 
 
+@cli.group(short_help='Image Related Commands')
+def image():
+    """
+    Use The Subcommands under this command to manipulate Snapshots
```
Comment (djfinn14) [inline ims/cli/cli.py]: Since we have moved toward the direction where this is not going to be in core BMI in the long run, but instead would more likely be a separate service, I renamed the group to just be "script" so it will be easier to break it out of the codebase.

**Comment by chemistry-sourabh** [comment]
@djfinn14 could you please comment on the comments that you solved with your new commit so that we can have a look.

File: ims/cli/cli.py [hunk 7]
```diff
@@ -87,6 +88,7 @@ def provision(project, node, img, network, nic):
             constants.NIC_PARAMETER: nic}
     res = requests.put(_url + "provision/", data=data,
                        auth=(_username, _password))
+    click.echo(res.url)
```
Comment (djfinn14) [inline ims/cli/cli.py]: Mistake, fixed.

File: ims/cli/cli.py (hunk 2 above)
Comment (djfinn14) [inline ims/cli/cli.py]: fixed

File: ims/cli/cli.py (hunk 6 above)
Comment (djfinn14) [inline ims/cli/cli.py]: changes made

File: ims/cli/cli.py [hunk 8]
```diff
@@ -671,5 +673,47 @@ def download():
...
+    """
+    Use The Subcommands under this command to manipulate Snapshots
+    """
+    pass
+
+
+@image.command(name='run_script', help='Run script within BMI')
+@click.argument(constants.PROJECT_PARAMETER)
+@click.argument(constants.IMAGE_NAME_PARAMETER)
+@click.argument(constants.SCRIPT_NAME_PARAMETER)
```
Comment (djfinn14) [inline ims/cli/cli.py]: fixed

File: ims/common/constants.py (hunk 1 above)
Comment (djfinn14) [inline ims/common/constants.py]: fixed.

File: ims/einstein/operations.py [hunk 9]
```diff
@@ -635,3 +639,31 @@ def remake_mappings(self):
             logger.exception('')
         except NotImplementedError:
             pass
+
+    @log
+    def run_script(self, img, script):
+        try:
+            if not self.is_admin:
+                raise AuthorizationFailedException()
+            directory = "/tmp/script/"
```
Comment (djfinn14) [inline ims/einstein/operations.py]: made the change.

File: ims/einstein/operations.py (hunk 3 above)
Comment (djfinn14) [inline ims/einstein/operations.py]: fixed.

File: ims/einstein/operations.py (hunk 5 above)
Comment (djfinn14) [inline ims/einstein/operations.py]: made the change.

**Comment by naved001** [comment]
@djfinn14 ran this for me and it works as expected.

**Comment by naved001** [comment]
The coverage has decreased because we didn't write tests for the new code. But we are making an exception here since it is for secure cloud paper.

I am gonna go ahead and merge this.
